dbms/src/main/java/org/polypheny/db/routing/strategies/RoutingPlanSelector.java

```diff
@@ -0,0 +1,235 @@
+
+
+/**
+ * Class which is responsible for cost calculation and plan selection.
+ */
+@Slf4j
+public class RoutingPlanSelector {
+
+    public RoutingPlan selectPlanBasedOnCosts(
+            List<? extends RoutingPlan> routingPlans,
+            List<RelOptCost> approximatedCosts,
+            Statement statement ) {
+        // 0 = only consider pre costs
+        // 1 = only consider post costs
+        // [0,1] = get normalized pre and post costs and multiply by ratio
+        final double ratioPre = 1 - RoutingManager.PRE_COST_POST_COST_RATIO.getDouble();
+        final double ratioPost = RoutingManager.PRE_COST_POST_COST_RATIO.getDouble();
+        final int n = routingPlans.size();
+
+        // Calculate pre-costs or set them to 0 for all entries.
+        boolean calcPreCosts = Math.abs( ratioPre ) >= RelOptUtil.EPSILON; // <=0
+        List<Double> preCosts = calcPreCosts ? normalizeApproximatedCosts( approximatedCosts ) : Collections.nCopies( n, 0.0 );
+
+        // Calculate post-costs or set them to 0 for all entries.
+        // If calculation is needed, get icarus original costs for printing debug output.
+        boolean calcPostCosts = Math.abs( ratioPost ) >= RelOptUtil.EPSILON; // > 0
+        List<Double> icarusCosts;
+        List<Double> postCosts;
+        List<Double> percentageCosts = null;
+        if ( calcPostCosts ) {
+            Pair<List<Double>, List<Double>> icarusResult = calculateIcarusPostCosts( routingPlans );
+            icarusCosts = icarusResult.right;
+            postCosts = icarusResult.left;
+        } else {
+            postCosts = Collections.nCopies( n, 0.0 );
+            icarusCosts = null;
+        }
+
+        // Get effective costs
+        RoutingPlan result = null;
+        List<Double> effectiveCosts = IntStream.rangeClosed( 0, n - 1 )
+                .mapToDouble( i -> (preCosts.get( i ) * ratioPre) + (postCosts.get( i ) * ratioPost) )
+                .boxed()
+                .collect( Collectors.toList() );
+
+        // Get plan in regard to the active strategy
+        if ( RoutingManager.PLAN_SELECTION_STRATEGY.getEnum() == RouterPlanSelectionStrategy.BEST ) {
+            result = this.selectBestPlan( routingPlans, effectiveCosts );
+        } else if ( RoutingManager.PLAN_SELECTION_STRATEGY.getEnum() == RouterPlanSelectionStrategy.PROBABILITY ) {
+            Pair<RoutingPlan, List<Double>> percentageResult = this.selectPlanFromProbability( routingPlans, effectiveCosts );
+            result = percentageResult.left;
+            percentageCosts = percentageResult.right;
+        }
+
+        if ( result == null ) {
+            throw new RuntimeException( "Plan selection strategy not found, result still null." );
+        }
+
+        if ( statement.getTransaction().isAnalyze() ) {
+            UiRoutingPageUtil.addRoutingAndPlanPage(
+                    approximatedCosts,
+                    preCosts,
+                    postCosts,
+                    icarusCosts,
```

Comment:
What is the difference between postCosts and icarusCosts?

Reply:
The naming in the class is a little bit a mess, I guess the term icarus should be removed completely.
But in general, postCosts are probabilities and icarusCosts are absolute measured times.